pkg/gcpclient/client.go

```diff
@@ -265,6 +305,392 @@ func (c *gcpClient) StartInstance(instance *compute.Instance) error {
+	}
+	return nil
+}
+
+func (c *gcpClient) GetAddress(addressName string, region string) (*compute.Address, error) {
+	ctx, cancel := contextWithTimeout(context.TODO())
+	defer cancel()
+
+	return c.computeClient.Addresses.Get(c.projectName, region, addressName).Context(ctx).Do()
+}
+
+// ListAddressesOptions are the options for listing compute images.
+type ListAddressesOptions struct {
+	MaxResults int64
+	PageToken  string
+	Filter     string
+}
+
+func (c *gcpClient) ListAddresses(region string, opts ListAddressesOptions) (*compute.AddressList, error) {
+	ctx, cancel := contextWithTimeout(context.TODO())
+	defer cancel()
+
+	call := c.computeClient.Addresses.List(c.projectName, region).Context(ctx).Filter(opts.Filter)
+
+	if opts.MaxResults > 0 {
+		call.MaxResults(opts.MaxResults)
+	}
+	if opts.PageToken != "" {
+		call.PageToken(opts.PageToken)
+	}
+
+	return call.Do()
+}
+
+func (c *gcpClient) CreateSubnet(
+	subnet string,
+	ipCidr string,
+	network string,
+	region string,
+	purpose string) (*compute.Subnetwork, error) {
+
+	ctx, cancel := contextWithTimeout(context.TODO())
+	defer cancel()
+
+	newSubnet := &compute.Subnetwork{
+		Name:        subnet,
+		Description: "",
+		IpCidrRange: ipCidr,
+		Network:     network,
+		Purpose:     purpose,
+		Region:      region,
+	}
+
+	op, err := c.computeClient.Subnetworks.Insert(c.projectName, region, newSubnet).Context(ctx).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	done, err := c.waitforRegionOperation(op, region)
+	if err != nil || !done {
+		return nil, err
+	}
+
+	return c.computeClient.Subnetworks.Get(c.projectName, region, subnet).Context(ctx).Do()
+}
+
+func (c *gcpClient) DeleteSubnet(subnet string, region string) error {
+	ctx, cancel := contextWithTimeout(context.TODO())
+	defer cancel()
+
+	op, err := c.computeClient.Subnetworks.Delete(c.projectName, region, subnet).Context(ctx).Do()
+	if err != nil {
+		return err
+	}
+
+	done, err := c.waitforRegionOperation(op, region)
+	if err != nil {
+		return err
+	}
+	if !done {
+		return errors.New("timed out waiting on deletion subnet")
+	}
+	return nil
+}
+
+func (c *gcpClient) GetSubnet(subnet string, region string) (*compute.Subnetwork, error) {
+	ctx, cancel := contextWithTimeout(context.TODO())
+	defer cancel()
+
+	return c.computeClient.Subnetworks.Get(c.projectName, region, subnet).Context(ctx).Do()
+}
+
+func (c *gcpClient) GetProjectName() string {
+	return c.projectName
+}
+
+func (c *gcpClient) waitforGlobalOperation(op *compute.Operation) (bool, error) {
+	// See if the op was already done
+	if op.Status == "DONE" {
+		if op.Error != nil && len(op.Error.Errors) > 0 {
+			return true, fmt.Errorf("operation completed with error: %s", op.Error.Errors[0].Code)
+		}
+		return true, nil
+	}
+
+	ctx, cancel := contextWithTimeout(context.TODO())
+	defer cancel()
+
+	// Wait for the op to finish (up to 2 minutes)
+	op, err := c.computeClient.GlobalOperations.Wait(c.projectName, op.Name).Context(ctx).Do()
```

Comment:
Will these block the main reconcile thread?

Reply:
Yes, similar to how awsprivatelink controller blocks the reconcile thread on [waitForState](https://github.com/openshift/hive/blob/f0759b0ad10b8f12ead2fdeead63bbadda7ff245/pkg/controller/awsprivatelink/awsprivatelink_controller.go#L630-L645). The goal is to have the threads wait until GCP API acknowledges the creation/deletion of objects before continuing. Otherwise, some things were returning here without immediate error. And then they error later and get removed. As a result, future steps also error out because they are attempting to use cloud resources that don't actually exist. By waiting here (which is a max of 2 minutes per the API docs), we avoid that in most cases. As a result, we give an error on why the thing wasn't created instead of a future error on another dependent resource.